### PR TITLE
Time filtering via OGR

### PR DESCRIPTION
**Pull request by SoerenHoffstedt**
Changed `initTimeReading` and `readTimeIntoCollection` in `OGRSourceUtil`.
When the time specification is `start+end` and the time attributes are of the OGRDataType `Date` or `DateTime` the filtering can be done with the `SetAttributeFilter` method of the `OGRLayer`.

Creating this filter can be forced even though the OGRDataTypes are wrong. This is useful for WFS requests. WFS will not report Date/DateTime data types correctly to OGR, because it sends the data as GML. But if the time attributes are Date or DateTime on the WFS server the filter will still work, because the filtering of `SetAttributeFilter` will be done on the server. But when the filter was applied falsely the request will return no features at all, because they will all be filtered.

### CODE REVIEW

File: src/util/ogr_source_util.cpp [hunk 1]
```diff
@@ -447,41 +449,125 @@ bool OGRSourceUtil::readAttributesIntoCollection(AttributeArrays &attributeArray
...
+	rect_time_String_1 += "\'";
+
+	std::string rect_time_string_2 = "\'";
+	rect_time_string_2 += rect.toIsoString(rect.t2);
+	rect_time_string_2 += "\'";
+
+	std::stringstream filter_stream;
+    // Create SQL like comparison to check if attributes validity intersects with the query rectangles times:
+    // (feature.t1 >= rect.t1 || feature.t2 >= rect.t1) && (feature.t1 <= rect.t2 || feature.t2 <= rect.t2)
+    filter_stream << "( " << time1Name << " >= " << rect_time_String_1;
```
Comment (michaelmattig) [inline src/util/ogr_source_util.cpp]: is it safe to just use the user specified values here or do we need to sanitize the input?

File: src/util/ogr_source_util.cpp [hunk 2]
```diff
@@ -533,13 +619,55 @@ bool OGRSourceUtil::readTimeIntoCollection(const QueryRectangle &rect, OGRFeatur
...
+    else if(time_zone == 1) //localtime
+    {
+		// Because we don't know when this happens and what it is supposed to mean in this context
+        // (local computer time, or the local time where the data was recorded [what we don't know]) we throw an exception for now.
+		throw OperatorException("OGRSource: Unexpected time zone value from the time attribute: \"localtime\"");
+    }
+    else
+    {
+        // time zone is explicitly given, 100 <-> +00, 104 <-> +01; 108 <-> +02; 96 <-> -01, etc
+        int zone_hours = (time_zone - 100) / 4;
```
Comment (michaelmattig) [inline src/util/ogr_source_util.cpp]: what about fractional time zones? https://www.timeanddate.com/time/time-zones-interesting.html

File: src/util/ogr_source_util.cpp [hunk 3]
```diff
@@ -447,41 +449,125 @@ bool OGRSourceUtil::readAttributesIntoCollection(AttributeArrays &attributeArray
...
+	// Both fields have to available and of type DateTime/Date, because only START_END as TimeSpecification is supported.
+	// That's because I couldn't find a way to express in the sql like query (time_attribute + time_duration).
+	if(timeSpecification != TimeSpecification::START_END)
+        return false;
+
+    // can try to set the filter without the fields being of type Date/DateTime. But that is unsafe and
+    // can result in empty collections. So the filter could be set on string data fields and still be a good query.
+    // but both attributes still have to be present and be of a Date/DateTime structure (as a string).
+    bool hasSaveTypes = (time1Type == OGRFieldType::OFTDateTime || time1Type == OGRFieldType::OFTDate) &&
+                           (time2Type == OGRFieldType::OFTDateTime || time2Type == OGRFieldType::OFTDate);
```
Comment (michaelmattig) [inline src/util/ogr_source_util.cpp]: is it safe to have mixed date and datetime for start and end?

**Review by michaelmattig — COMMENTED** [review]
please auto format the file in addition to my comments

File: src/util/ogr_source_util.cpp (hunk 3 above)
Comment (SoerenHoffstedt) [reply src/util/ogr_source_util.cpp]: Yes, it is tested. Both types are always compared with DateTime Strings and reading the time later works the same for both types.

File: src/util/ogr_source_util.cpp (hunk 2 above)
Comment (SoerenHoffstedt) [reply src/util/ogr_source_util.cpp]: Support is added.

File: src/util/ogr_source_util.cpp (hunk 1 above)
Comment (SoerenHoffstedt) [reply src/util/ogr_source_util.cpp]: At this point we know that `time1Name`/`time2Name` are existing attributes. We also know they have a supported data type (or are Strings if the filtering is forced). So I think this is fine.

**Pull request merged**